ggml-cuda.cu

```diff
@@ -173,6 +206,40 @@ static __global__ void dequantize_block_q8_0(const void * vx, float * y) {
     }
 }
 
+template <int block_size, int qk, dequantize_kernel_t dequantize_kernel> static __global__ void dequantize_mul_mat_vec(const void * vx, const float * y, float * dst, const int ncols) {
+    const int row = blockIdx.x;
+    const int tid = threadIdx.x;
+
+    __shared__ float tmp[block_size]; // separate sum for each thread
+    tmp[tid] = 0;
+
+    for (int i = 0; i < ncols/block_size; i += 2) {
+        const int col = i*block_size + 2*tid;
+        const int ib = (row*ncols + col)/qk; // block index
+        const int iqs = (col%qk)/2; // quant index
+        const int iybs = col - col%qk; // y block start index
+
+        // dequantize
+        float v0, v1;
+        dequantize_kernel(vx, ib, iqs, v0, v1);
+
+        // matrix multiplication
+        tmp[tid] += v0 * y[iybs + iqs + 0];
+        tmp[tid] += v1 * y[iybs + iqs + qk/2];
+    }
+
+    // sum up partial sums and write back result
+    for (int s=block_size/2; s>0; s>>=1) {
```

Comment:
I believe you need a `__syncthreads();` call before the start of this loop

Reply:
Yeah, good catch.

Reply:
I'll trust your expertise.

Reply:
Sorry, I should have explained why I think it's needed (not good to trust anyone without explanation). My reasoning:
  * if there were more than one warp, then certainly `__syncthreads()` would be needed to ensure the faster warps don't start reducing before the slower ones are finished with their computations
  * even if there is only one warp (the way it is now), some sort of synchronisation is still needed on Volta and later architectures, since they have [independent thread scheduling](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#independent-thread-scheduling). The docs explicitly say "assumptions that code is executed in lockstep or that reads/writes *[from/to shared memory – my clarification]* from separate threads are visible across a warp without synchronization are invalid". They recommend `__syncwarp()` for intra-warp synchronization, but I think `__syncthreads()` should be also perfectly legal (since it is stronger).